src/sessions/p2p_session.rs

```diff
@@ -842,4 +862,49 @@ impl<T: Config> P2PSession<T> {
             self.event_queue.pop_front();
         }
     }
+
+    fn compare_local_checksums_against_peers(&mut self) {
+        let current = self.current_frame();
+        if current % CHECKSUM_REPORT_INTERVAL == 0 && current > self.max_prediction as i32 {
+            for (_, remote) in &self.player_reg.remotes {
+                for (remote_frame, remote_checksum) in remote.checksum_history() {
+                    for (local_frame, local_checksum) in &self.checksum_history {
```

Comment:
Can we avoid the n*n complexity here? If the checksums histories are sorted, maybe they could be traversed together? Or use a hashmap instead?

Reply:
my previous implementation before using VeqDeques used HashMaps but the housekeeping and tracking in my first implementation was hard to keep up with. i will try to make use of them again. or maybe there's a data structure that might works better 🤔.

Reply:
Right now the history size is 128, so it's num_players*size*size iterations, so it's currently at 65,536 iterations with 4 players. Maybe it's negligible, but perhaps @gschup would like to make the call whether it's worth it?

Just reducing the history size to something like 32 would make it much more managable, though...

Reply:
fwiw, rebuilding a hash set from scratch every time this function is called would also be faster than the current nested loops.

so maybe that's some middle ground to consider?

Reply:
Actually, once we've compared with a remote checksum, we don't really need it anymore? Maybe we can just remove it so we don't iterate over the same entries again and again?